test/Sema/pattern_matching_optional_promotion.swift.gyb

```diff
@@ -0,0 +1,205 @@
+        // CHECK:     (case_label_item
+        // CHECK-NEXT:  (pattern_enum_element type='E??' E??.some
+        // CHECK-NEXT:    (pattern_paren type='(E?)'
+        // CHECK-NEXT:      (pattern_optional_some implicit type='(E?)'
+        // CHECK-NEXT:        (pattern_enum_element type='E' E.a
+      }
+    }
+  }
+
+  do {
+    ${test_case_introducer} {
+      switch oe {
+        case nil: ()
+        // FIXME: '.a' is OK but 'E.a' is not?
+        case E.a: () // expected-error {{enum case 'a' is not a member of type 'E?'}}
+      }
+    }
+  }
+
+% end
+}
+
+// Tuple patterns.
+do {
+  var ot: (E, E)?
+  var oot: (E, E)??
+  var tot: ((E, E)?, E)
+
+% for mode in modes:
+%   test_case_introducer = get_test_case_introducer(mode)
+  do {
+    ${test_case_introducer} {
+      // FIXME: Works in a 'switch' but not an 'if'.
+      if case (.a, .a) = ot {}
+      // expected-error@-1 {{value of optional type '(E, E)?' must be unwrapped to a value of type '(E, E)'}}
+      // expected-note@-2 {{coalesce}}
+      // expected-note@-3 {{force-unwrap}}
+    }
+  }
+
+  do {
+    ${test_case_introducer} {
+      switch ot {
+        case nil: ()
+        case (.a, .a): () // CHECK: (case_stmt range=[{{.+}}.swift:[[@LINE]]
+        // CHECK:      (case_label_item
+        // CHECK-NEXT: (pattern_optional_some implicit type='(E, E)?'
+        // CHECK-NEXT:   (pattern_tuple type='(E, E)'
+      }
+      switch ot {
+        case nil: ()
+        case let (x, y): () // CHECK: (case_stmt range=[{{.+}}.swift:[[@LINE]]
+        // CHECK:      (case_label_item
+        // CHECK-NEXT:   (pattern_let type='(E, E)?'
+        // CHECK-NEXT:     (pattern_optional_some implicit type='(E, E)?'
+        // CHECK-NEXT:        (pattern_tuple type='(E, E)'
+      }
+      switch oot {
+        case nil, .some(nil): ()
+        case (.a, .a): () // CHECK: (case_stmt range=[{{.+}}.swift:[[@LINE]]
+        // CHECK:      (case_label_item
+        // CHECK-NEXT: (pattern_optional_some implicit type='(E, E)??'
+        // CHECK-NEXT:   (pattern_optional_some implicit type='(E, E)?'
+        // CHECK-NEXT:     (pattern_tuple type='(E, E)'
+      }
+      switch oot {
+        case nil, .some(nil): ()
+        case (.a, .a)?: () // CHECK: (case_stmt range=[{{.+}}.swift:[[@LINE]]
+        // CHECK:      (case_label_item
+        // CHECK-NEXT: (pattern_optional_some type='(E, E)??'
+        // CHECK-NEXT:   (pattern_optional_some implicit type='(E, E)?'
+        // CHECK-NEXT:     (pattern_tuple type='(E, E)'
+      }
+      switch tot {
+        case (nil, .a): ()
+        case ((.a, .a), .a): () // CHECK: (case_stmt range=[{{.+}}.swift:[[@LINE]]
+        // CHECK:      (case_label_item
+        // CHECK-NEXT:   (pattern_tuple type='((E, E)?, E)'
+        // CHECK-NEXT:     (pattern_optional_some implicit type='(E, E)?'
+        // CHECK-NEXT:       (pattern_tuple type='(E, E)'
+      }
+    }
+  }
+
+% end
```

Comment:
Thanks for adding such thorough tests! Would you mind also adding some for `if case` and `for case`?